### ISSUE TITLE
parents and degrees of PermGroup

### ISSUE THREAD

**Issue by fieker** [oscar-system/Oscar.jl#894]
```
julia> G = symmetric_group(2)
Sym( [ 1 .. 2 ] )

julia> degree(G)
2

julia> G = symmetric_group(1)
Group(())

julia> degree(G)
0
```
This should be 1.

The next:
```
julia> G = symmetric_group(3)
Sym( [ 1 .. 3 ] )

julia> U = sub(G, [G([1,2,3])])[1]
Group(())
julia> U^U([1,2,3])
Group(())

julia> ans([1,2,3])
ERROR: ArgumentError: the element does not embed in the group
Stacktrace:
 [1] (::PermGroup)(L::Vector{Int64})
   @ Oscar ~/.julia/dev/Oscar/src/Groups/perm.jl:162
 [2] top-level scope
   @ REPL[85]:1
```
so the conjugation looses the degree...

**Comment by ThomasBreuer** [comment]
The first problem is easy to fix.

For the second one, we have to decide what we want.
When I conjugate a permutation group G with a permutation x, the degrees of G and x may be different.
If we want to support this general situation and if we want that the degree of the result depends only on the degrees of G and x (and not on the moved points of the result) then the only sensible definition seems to be the maximum of the two degrees.

(I will provide a pull request.)

**Comment by fieker** [comment]
On Thu, Dec 16, 2021 at 03:24:02AM -0800, Thomas Breuer wrote:
> The first problem is easy to fix.
> 
> For the second one, we have to decide what we want.
> When I conjugate a permutation group G with a permutation x, the degrees of G and x may be different.
In Gap: yes.
In Oscar: if the degrees differ, the element and the group are not
compatible. In my opinion, this should then error

**Comment by thofma** [comment]
Agree with Claus here.

**Comment by ThomasBreuer** [comment]
O. k., let us change the behaviour such that we throw an error whenever incompatible groups are entered.
This should then also hold for the multiplication and conjugation of group elements.

The current approach via `_maxgroup` can then be abolished.
Anyway it leads to effects which are difficult to explain, for example this one:
```
julia> g = sylow_subgroup(symmetric_group(5), 3)[1]
Group([ (1,2,3) ])

julia> h = symmetric_group(4)
Sym( [ 1 .. 4 ] )

julia> degree(g)
5

julia> degree(h)
4

julia> m = Oscar._common_parent_group(g, h);;  m == h   # w.r.t. inclusion, h contains g
true

julia> degree(parent(gen(g, 1) * gen(h, 1)))    # and here the degree goes down
4
```

**Comment by fieker** [comment]
what is happening here? The reason I am asking is that this is a blocking error to deal with some Galois problems.

**Comment by fingolfin** [comment]
We just discussed this. The current plan then is: indeed require equal degrees, otherwise throw error; but then also add an "easy" way to coerce a given group into a different degree. E.g. (pseudo code):
```julia
julia> G = symmetric_group(10); H = symmetric_group(5); H_new = G(H);

julia> degree(G)
10

julia> degree(H)
5

julia> degree(H_new)
10
```

**Comment by fingolfin** [comment]
The two issues originally reported here are resolved.

The odd example with `_maxgroup` resp. `_common_parent_group` described by @ThomasBreuer still is as it was. 

So we still need to change the behaviour such that we throw an error whenever incompatible groups are entered.

**Comment by ThomasBreuer** [comment]
O.k., let us get rid of `_common_parent_group`.
I will provide a pull request for that.

**Comment by fingolfin** [comment]
On monday Thomas and me discussed the possibility of unifying this across group types, i.e., perm, (sub)fp, (sub)pc, matrix ... for handling the "common parent situation".

The plan is this:
- for each type there is a clear "grand parent" / "top parent" / "universal parent" / your-name-here group: 
  - `symmetric_group(n)` for a permutation or or permutation group of degree $n$,
  - `GL(n,r)` for a `n\times n` matrix group over some ring $R$,
  - some `PcGroup` for any `SubPcGroup` 
  - some `FpGroup` for any `SubFpGroup`
  - `FpGroup` and `PcGroup` instances already are "top dogs"

Multiplying two group elements shall only be supported if the two elements have "the same" grand parent. The resulting product will have that grand parent as its parent (unless the two initial elements have the same parent, in which case the product can also use that). In the future this may or may not be refined to use a "partial poset of subgroups" of the grandparent.

(Somewhat related: for permutations and permutation groups we should cache the `symmetric_group(n)` used as parent somehow, see issue #129)

**Comment by thofma** [comment]
Just to check that I understand it correctly. Assume I have two subgroups `H, K` of `GL(2, 3)` and an elements `x`, `y` with parent equal to `H` and `K` respectively. Will `x * y` throw an error or will it construct the element with parent equal to `GL(2, 3)`? What is the current behaviour and what is the plan?

**Comment by fingolfin** [comment]
```
julia> G = GL(2,3)
GL(2,3)

julia> H = SL(2,3)
SL(2,3)

julia> K = GO(1, 2,3)
GO+(2,3)

julia> H[1] in K
false

julia> K[2] in H
false

julia> parent(H[1])
SL(2,3)

julia> parent(K[2])
GO+(2,3)

julia> g = H[1] * K[2]
[1   1]
[1   0]

julia> parent(g)
GL(2,3)
```

This is current behavior and for now the plan would be for it to stay this way, but we are always open to hear additional viewpoints and arguments.

**Comment by ThomasBreuer** [comment]
My point of view is:
For any Oscar group that is handled via a group in GAP, we can delegate the multiplication of two group elements to GAP. For those situations where GAP returns a group element, Oscar has to decide what happens. Either a parent object must be chosen/constructed, and the GAP result gets wrapped into an Oscar group element, or the multiplication is regarded as not allowed for some reason (as shall be done for elements of permutation groups).

The currently implemented rules in this respect are not consistent:
For matrix groups, we have the "assign to GL" strategy shown above, but generically there is a method that compares whether one parent is contained in the other parent, and if yes then the bigger parent is chosen. This work in many cases, but not always. Here is an obvious example.
```
julia> g = small_group(24, 12)
Pc group of order 24

julia> s2 = sylow_subgroup(g, 2)[1]
Sub-pc group of order 8

julia> s3 = sylow_subgroup(g, 3)[1]
Sub-pc group of order 3

julia> s2[1] * s3[1]
ERROR: Not yet implemented
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:35
 [2] _common_parent_group
   @ /export/home/sam/julia/packages/Oscar.jl/src/Groups/GAPGroups.jl:231 [inlined]
 [3] _prod(x::SubPcGroupElem, y::SubPcGroupElem)
   @ Oscar /export/home/sam/julia/packages/Oscar.jl/src/Groups/GAPGroups.jl:240
```
The idea outlined above would create a group element with parent `g`. This would be analogous to the "assign to GL" rule, since `g` is the largest possible group into which the element can be naturally embedded.

**Comment by thofma** [comment]
1. Can constructing the "natural" overgroup be expensive?
2. Is this only for group element operations?
3. Can you summarize what happended between
> O.k., let us get rid of _common_parent_group.

  and

> On monday Thomas and me discussed the possibility of unifying this across group types

i.e., why do we want to keep the "grandpartent" concept? Because we have it for matrix groups and are stuck with it forever? Or is there another reason?

**Comment by ThomasBreuer** [comment]
1. For `SubFpGroup` and `SubPcGroup`, we store the "full group" already. For matrix groups as well as for permutation groups, some caching will be needed.
2. Currently the code in question is called only in the arithmetic operations of group elements, but there are situations such as certain closure operations of subgroups which are supported in GAP but currently not in Oscar.
3. I am simply not sure what is a really good solution ...
